src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,159 @@
+                    if (mappingResult != null)
+                    {
+                        if (mappingResult.Symbol is INamedTypeSymbol namedTypeSymbol)
+                        {
+                            // Find its baseTypes and subTypes.
+                            await AddInheritanceMemberItemsForNamedTypeAsync(
+                                mappingResult.Project.Solution,
+                                sourceText,
+                                memberDeclarationNode,
+                                namedTypeSymbol, builder, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        if (mappingResult.Symbol is IEventSymbol or IPropertySymbol || mappingResult.Symbol.IsOrdinaryMethod())
+                        {
+                            // Find the implementing/implemented/overriden/overriding members
+                            await AddInheritanceMemberItemsForTypeMembersAsync(
+                                mappingResult.Project.Solution,
+                                sourceText,
+                                memberDeclarationNode,
+                                mappingResult.Symbol,
+                                builder,
+                                cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static async Task AddInheritanceMemberItemsForNamedTypeAsync(
+            Solution solution,
+            SourceText sourceText,
+            SyntaxNode declarationNode,
+            INamedTypeSymbol memberSymbol,
+            ArrayBuilder<InheritanceMemberItem> builder,
+            CancellationToken cancellationToken)
+        {
+            // Get all base types
+            var allBaseTypes = BaseTypeFinder.FindBaseTypesAndInterfaces(memberSymbol);
+
+            // Filter out System.Object. (otherwise margin would be shown for all classes)
+            var baseTypes = allBaseTypes
+                .WhereAsArray(type =>
+                    !(type is ITypeSymbol { SpecialType: SpecialType.System_Object }));
+
+            // Get all derived types
+            var derivedTypes = await GetDerivedTypesAndImplementationsAsync(
+                solution,
+                memberSymbol,
+                cancellationToken).ConfigureAwait(false);
+
+            if (!(baseTypes.IsEmpty && derivedTypes.IsEmpty))
+            {
+                var lineNumber = sourceText.Lines.GetLineFromPosition(declarationNode.SpanStart).LineNumber;
+                var item = await CreateInheritanceMemberInfoAsync(
+                    solution,
+                    memberSymbol,
+                    lineNumber,
+                    baseSymbols: baseTypes,
+                    derivedTypesSymbols: derivedTypes.OfType<ISymbol>().ToImmutableArray(),
```

Comment:
💡 Use `CastArray<ISymbol>()` or `SelectAsArray<INamedTypeSymbol, ISymbol>(static symbol => symbol)`

```suggestion
                    derivedTypesSymbols: derivedTypes.CastArray<ISymbol>(),
```

```suggestion
                    derivedTypesSymbols: derivedTypes.SelectAsArray<INamedTypeSymbol, ISymbol>(static symbol => symbol),
```

Reply:
Great to know the first helper. : )